File: go/internal/orchestrator/sharding.go
// Package orchestrator — ShardPlanner groups files into token-budget shards.
package orchestrator

import (
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/unrealandychan/close-wiki/internal/models"
)

const (
	bytesPerToken      = 4     // ~4 bytes per token
	defaultTokenBudget = 40000 // max tokens per shard; aligned with Python runtime
)

// resolveTokenBudget returns the effective token budget.
// Checks REKIPEDIA_SHARD_TOKEN_BUDGET env var first; falls back to defaultTokenBudget.
func resolveTokenBudget(override int) int {
	if override > 0 {
		return override
	}
	if v := os.Getenv("REKIPEDIA_SHARD_TOKEN_BUDGET"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 {
			return n
		}
	}
	return defaultTokenBudget
}

// ShardPlanner groups FileManifest entries into shards that fit within a token budget.
type ShardPlanner struct {
	budget int
}

// NewShardPlanner creates a ShardPlanner with the given token budget.
// If budget ≤ 0, checks REKIPEDIA_SHARD_TOKEN_BUDGET env var, then falls back to 40 000.
func NewShardPlanner(budget int) *ShardPlanner {
	return &ShardPlanner{budget: resolveTokenBudget(budget)}
}

// Plan groups files by top-level directory and splits groups that exceed the budget.
func (sp *ShardPlanner) Plan(files []models.FileManifest) []models.Shard {
	if len(files) == 0 {
		return nil
	}

	// Group by top-level directory
	groups := make(map[string][]models.FileManifest)
	var order []string
	seen := make(map[string]bool)
	for _, f := range files {
		top := topLevelDir(f.Path)
		if !seen[top] {
			order = append(order, top)
			seen[top] = true
		}
		groups[top] = append(groups[top], f)
	}

	var shards []models.Shard
	for _, dir := range order {
		shards = append(shards, sp.splitGroup(dir, groups[dir])...)
	}
	return shards
}

func (sp *ShardPlanner) splitGroup(groupDir string, files []models.FileManifest) []models.Shard {
	var shards []models.Shard
	var bucket []models.FileManifest
	bucketTokens := 0
	bucketIdx := 0

	for _, f := range files {
		fileTokens := fileTokenEstimate(f.SizeBytes)
		if len(bucket) > 0 && bucketTokens+fileTokens > sp.budget {
			shards = append(shards, makeShard(groupDir, bucketIdx, bucket))
			bucket = nil
			bucketTokens = 0
			bucketIdx++
		}
		bucket = append(bucket, f)
		bucketTokens += fileTokens
	}
	if len(bucket) > 0 {
		shards = append(shards, makeShard(groupDir, bucketIdx, bucket))
	}
	return shards
}

func makeShard(groupDir string, idx int, files []models.FileManifest) models.Shard {
	id := groupDir
	if idx > 0 {
		id = groupDir + "#" + string(rune('0'+idx))
	}
	return models.Shard{
		ShardID: id,
		Root:    groupDir,
		Files:   files,
	}
}

func topLevelDir(path string) string {
	path = filepath.ToSlash(path)
	parts := strings.SplitN(path, "/", 2)
	if len(parts) > 1 {
		return parts[0]
	}
	return "."
}

func fileTokenEstimate(sizeBytes int64) int {
	tokens := int(sizeBytes) / bytesPerToken
	if tokens < 1 {
		return 1
	}
	return tokens
}

File: src/rekipedia/orchestrator/sharding.py
"""Shard planner: partition FileManifest lists into token-budget shards."""
from __future__ import annotations

from pathlib import Path

from rekipedia.models.contracts import FileManifest, LLMConfig, Shard

# Rough tokens-per-byte ratio (4 chars ≈ 1 token for code)
_BYTES_PER_TOKEN = 4
# Max tokens per shard before a new shard is started
_DEFAULT_TOKEN_BUDGET = 12_000


class ShardPlanner:
    """Group files into shards that fit within a token budget.

    Files are first grouped by top-level directory.  If a group exceeds the
    budget it is split into consecutive sub-groups.
    """

    def __init__(self, token_budget: int = _DEFAULT_TOKEN_BUDGET) -> None:
        self._budget = token_budget

    def plan(
        self,
        files: list[FileManifest],
        llm: LLMConfig | None = None,
    ) -> list[Shard]:
        if not files:
            return []

        llm = llm or LLMConfig()

        # Group by top-level directory (or "." for root files)
        groups: dict[str, list[FileManifest]] = {}
        for f in files:
            top = Path(f.path).parts[0] if "/" in f.path or "\\" in f.path else "."
            groups.setdefault(top, []).append(f)

        shards: list[Shard] = []
        for group_dir, group_files in groups.items():
            shards.extend(self._split_group(group_dir, group_files, llm))

        return shards

    def _split_group(
        self,
        group_dir: str,
        files: list[FileManifest],
        llm: LLMConfig,
    ) -> list[Shard]:
        shards: list[Shard] = []
        bucket: list[FileManifest] = []
        bucket_tokens = 0
        bucket_idx = 0

        for f in files:
            file_tokens = max(1, f.size_bytes // _BYTES_PER_TOKEN)
            if bucket and bucket_tokens + file_tokens > self._budget:
                shards.append(_make_shard(group_dir, bucket_idx, bucket, llm))
                bucket = []
                bucket_tokens = 0
                bucket_idx += 1
            bucket.append(f)
            bucket_tokens += file_tokens

        if bucket:
            shards.append(_make_shard(group_dir, bucket_idx, bucket, llm))

        return shards


def _make_shard(
    group_dir: str, idx: int, files: list[FileManifest], llm: LLMConfig
) -> Shard:
    shard_id = group_dir if idx == 0 else f"{group_dir}#{idx}"
    return Shard(shard_id=shard_id, root=group_dir, files=files, llm=llm)
